test/controllers/orders_controller_test.rb

```diff
@@ -0,0 +1,274 @@
+
+      old_order_count = Order.count
+
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :bad_request
+      @order.reload
+      @order.name.wont_equal @order_data[:name]
+      Order.count.must_equal old_order_count
+
+
+      # with future date more than 5 years
+
+      @order_data[:ccexpiration] = Date.today + 6.years
+
+      old_order_count = Order.count
+
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :bad_request
+      @order.reload
+      @order.name.wont_equal @order_data[:name]
+      Order.count.must_equal old_order_count
+
+      #  with cvv lenght more than 4
+      @order_data[:cvv] = "09831"
+
+      old_order_count = Order.count
+
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :bad_request
+      @order.reload
+      @order.name.wont_equal @order_data[:name]
+      Order.count.must_equal old_order_count
+
+      # with ccnumber less than 16
+
+      @order_data[:ccnumber] = "13538202"
+
+      old_order_count = Order.count
+
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :bad_request
+      @order.reload
+      @order.name.wont_equal @order_data[:name]
+      Order.count.must_equal old_order_count
+
+      # with ccnumber more than 19
+
+      @order_data[:ccnumber] = "1234567890987654321"
+
+      old_order_count = Order.count
+
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :bad_request
+      @order.reload
+      @order.name.wont_equal @order_data[:name]
+      Order.count.must_equal old_order_count
+    end
+
+    it "won't update existing order with insufficient data" do
+      # without name
+      @order_data[:name] = nil
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :bad_request
+      @order.reload
+      @order.name.wont_equal @order_data[:name]
+
+
+      # without ccnumber
+      @order_data[:ccnumber] = nil
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :bad_request
+      @order.reload
+      @order.name.wont_equal @order_data[:name]
+
+
+      # without cvv
+      @order_data[:cvv] = nil
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :bad_request
+      @order.reload
+      @order.name.wont_equal @order_data[:name]
+
+
+      # without ccexpiration
+      @order_data[:ccexpiration] = nil
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :bad_request
+      @order.reload
+      @order.name.wont_equal @order_data[:name]
+
+      # without address
+      @order_data[:address] = nil
+      patch order_path(@order), params: { order: @order_data }
+
+      must_respond_with :bad_request
+      @order.reload
+      @order.name.wont_equal @order_data[:name]
+
+
+      # without zipcode
```

Comment:
This is the same fundamental test and with pieces of the data changed. Can you think of how you could consolidate to be able to use a loop rather than repeating the tests each time?